databricks/koalas/frame.py

```diff
@@ -3225,6 +3225,105 @@ def mask(self, cond, other=np.nan) -> "DataFrame":
+
+        The mode of a set of values is the value that appears most often.
+        It can be multiple values.
+
+        Parameters
+        ----------
+        axis : {0 or 'index', 1 or 'columns'}, default 0
+            The axis to iterate over while searching for the mode:
+
+            * 0 or 'index' : get mode of each column
+            * 1 or 'columns' : get mode of each row.
+
+        numeric_only : bool, default False
+            If True, only apply to numeric columns.
+        dropna : bool, default True
+            Don't consider counts of NaN/NaT.
+
+        Returns
+        -------
+        DataFrame
+            The modes of each column or row.
+
+        See Also
+        --------
+        Series.mode : Return the highest frequency value in a Series.
+        Series.value_counts : Return the counts of values in a Series.
+
+        Examples
+        --------
+        >>> kdf = ks.DataFrame(
+        ...     [("bird", 2, 2), ("mammal", 4, np.nan), ("arthropod", 8, 0), ("bird", 2, np.nan)],
+        ...     index=("falcon", "horse", "spider", "ostrich"),
+        ...     columns=("species", "legs", "wings"),
+        ... )
+        >>> kdf
+                   species  legs  wings
+        falcon        bird     2    2.0
+        horse       mammal     4    NaN
+        spider   arthropod     8    0.0
+        ostrich       bird     2    NaN
+
+        >>> kdf.mode().sort_values(by=list(kdf.columns)).reset_index(drop=True)
+          species  legs  wings
+        0    bird   2.0    0.0
+        1    None   NaN    2.0
+
+        >>> kdf.mode(dropna=False)
+          species  legs  wings
+        0    bird     2    NaN
+
+        >>> kdf.mode(numeric_only=True).sort_values(by='legs').reset_index(drop=True)
+           legs  wings
+        0   2.0    0.0
+        1   NaN    2.0
+        """
+        axis = validate_axis(axis)
+        assert axis == 0
+
+        data = self if not numeric_only else self._get_numeric_data()
+
+        new_scol = verify_temp_column_name(data._internal.spark_frame, "__row_index__")
+
+        def scol_mode(col):
+            if dropna:
+                sdf_dropna = data._internal.spark_frame.select(col).dropna()
+            else:
+                sdf_dropna = data._internal.spark_frame
+            count_df = sdf_dropna.groupBy(col).count()
+            most_value = count_df.orderBy("count", ascending=False).first()[1]
+            sdf_most_value = count_df.filter("count == {}".format(most_value)).select(col)
+
+            return sdf_most_value.withColumn(
+                new_scol, F.row_number().over(Window.orderBy(F.monotonically_increasing_id()))
+            )
+
+        new_sdf = None
+        for data_scol_name in data._internal.data_spark_column_names:
+            if new_sdf is None:
+                new_sdf = scol_mode(data_scol_name)
+            else:
+                new_sdf = new_sdf.join(scol_mode(data_scol_name), on=[new_scol], how="outer")
```

Comment:
I'm afraid there are so many joins as the number of columns. We want to avoid such cases.
I guess this is a similar case to what we did for `idxmax`/`idxmin`. https://github.com/databricks/koalas/pull/1054#discussion_r352971248

Reply:
Hm, I can't find a good condition to reduce the amount of data so far..

Reply:
Yeah, I see what you mean. In my current implementation, `scol_mode` will generate a new Spark frame per `data_scol_name`, and we combine these Spark frames along the way.

Would it be acceptable with a warning in the docstring showing this operations is expensive?